Add command to find a `kustomize` build base

matthewhughes-uw opened the pull request:
The current approach to finding a build base[1] is a bit too naive. It
just uses the names from `git-diff` and excludes deleted files, this was
intentional to get something simple running. This command should be
useable as-is in the current pipeline by replacing:

    git diff --name-only --diff-filter=d HEAD^1 HEAD

with:

    git diff --name-only --raw HEAD^1 HEAD | get-kustomize-build-base

This command aims to address this naivety by having explicit actions on
different types of changes, e.g. if we delete a `kustomization.yaml` we
should walk up the tree until we find another and declare that as the
build root.

This is another iteration with the goal of getting _something_ working,
and not necessarily perfect, some tradeoffs are noted in the README,
though additionally the command expects you to pass the `build-root`
(basically, root of the Git repo) rather than computing that itself. But
I think it's a sufficient improvement to warrant experimenting with it.

Some examples, running it on [2]

    kubernetes-manifests$ git diff --raw 280d134c02222dd43e9f3230f7378dd838692581~ 280d134c02222dd43e9f3230f7378dd838692581 | ../manifest-checkers/get-kustomize-build-base --build-root "$PWD" | sort --unique
    dev-aws/sys-mon/thanos-shared

And [3] (un-merged, so checked out at `7ae644208385efbeeac7ca7e34cb887f5df80cf7`)

    kubernetes-manifests$ git diff --raw "$(git merge-base master HEAD)" | ../manifest-checkers/get-kustomize-build-base --build-root "$PWD" | sort --unique
    dev-merit/dev-enablement
    dev-merit/kube-system

I also tested using a patched version of `kustomize-build-dirs` that
just prints the directories it would build, here's the patch:

    diff --git a/cmd/kustomize-build-dirs/main.go b/cmd/kustomize-build-dirs/main.go
    index 3d6bd69..880bd22 100644
    --- a/cmd/kustomize-build-dirs/main.go
    +++ b/cmd/kustomize-build-dirs/main.go
    @@ -74,6 +74,10 @@ func kustomizeBuildDirs(outDir string, doTruncateSecrets bool, filepaths []strin
            if err != nil {
                    return err
            }
    +       for _, root := range kustomizationRoots {
    +               fmt.Println(root)
    +       }
    +       return nil

            kustomizationRoots, err = removeComponentKustomizations(rootDir, kustomizationRoots)
            if err != nil { //go-cov:skip
    @@ -127,7 +131,7 @@ func findKustomizationRoots(root string, paths []string) ([]string, error) {
                    }

                    if _, exists := rootsMap[kustomizationRoot]; !exists {
    -                       fmt.Printf("Found kustomization build dir: %s\n", kustomizationRoot)
    +                       //fmt.Printf("Found kustomization build dir: %s\n", kustomizationRoot)
                            rootsMap[kustomizationRoot] = struct{}{}
                    }
            }

Via a script:

    #!/usr/bin/env bash

    set -o errexit -o pipefail -o nounset

    while read -r commit_sha
    do
        our_bases="$(git diff --raw "$commit_sha"^1 "$commit_sha" | ../manifest-checkers/get-kustomize-build-base --build-root "$PWD" | sort --unique)"
        their_bases="$(git diff --diff-filter=d --name-only "$commit_sha"^1 "$commit_sha" | xargs ../manifest-checkers/kustomize-build-dirs --out-dir /dev/null | sort)"

        if ! diff <(echo "$our_bases") <(echo "$their_bases")
        then
            echo "diff for $commit_sha"
        fi

    done < <(git log --format='%H' --max-count 100)

It reported two differences, both where a new `kustomization.yaml` was
added, i.e. old `kustomize-build dirs` wouldn't have built these, new
one would:

    1d0
    < prod-aws/energy-smart/fabricators
    diff for cfb45a6726d640c17926fdd3056dd58ad82db0f3
    1d0
    < dev-merit/partner-commission
    diff for eb130425af15096fa7984770295199ae745551d8

[1] https://github.com/utilitywarehouse/kubernetes-manifests/blob/d0a3f5eca5a47ae9f0a5cf42d1f6e8a9994fe7bb/.github/workflows/manifest-checks.yaml#L54
[2] https://github.com/utilitywarehouse/kubernetes-manifests/pull/92921
[3] https://github.com/utilitywarehouse/kubernetes-manifests/pull/98196

Ticket: DENA-842

MarcinGinszt commented on cmd/get-kustomize-build-base/main.go:
```diff
@@ -0,0 +1,141 @@
+	}
+}
+
+func runApp(ctx context.Context, out io.Writer, in io.Reader, args []string) error {
```
Why do we need this function? 
We are splitting the very basic main() function into three components.
If we are not solving any problem this way, wouldn't it be better to just use the convention from the repository?

matthewhughes-uw replied:
> Why do we need this function? We are splitting the very basic main() function into three components. If we are not solving any problem this way, wouldn't it be better to just use the convention from the repository?

I like my main to do two things:

* Call some other function
* Exit with a corresponding code from that function

I don't like having logic in `main` because It's hard to test: if I want to test anything with args I have to patch `os.Args`, and since I `os.Exit` I can't really test anything about the exit code. Additionally, since I want to `os.Exit` from main I can't expect any `defer`s to be run, so I don't want my main doing any additional setup etc. that needs any resource management since I might run into issues with this

MarcinGinszt commented on cmd/get-kustomize-build-base/README.md:
```diff
@@ -0,0 +1,24 @@
+# `get-kustomize-build-base`
```
Correct me if I'm wrong, but this component doesn't serve any purpose right now, right?
Because it replaces just part of `kustomize-build-dirs` logic. So, I understand, that with this PR you are just preparing the repository for actual changes.

MarcinGinszt replied:
If so, I think this should be included in the PR description. 
It would also be worth to include information about your planning next steps- how are you finally planning to use this piece of work?
Right now, I honestly don't know what are we supposed to review

matthewhughes-uw replied:
> Correct me if I'm wrong, but this component doesn't serve any purpose right now, right? Because it replaces just part of `kustomize-build-dirs` logic. So, I understand, that with this PR you are just preparing the repository for actual changes.

it can be slotted in just before the call to `kustomize-build-dirs`: instead of `git diff --name-only ... | xargs kustomize-build-dirs` we can `git diff --raw ... | get-kustomize-build-base | xargs kustomize-build-dirs`. `kustomize-build-dirs` still has logic to find a build base, but since the values we pass it should already be build bases it should just work™

MarcinGinszt replied:
> it can be slotted in just before the call to kustomize-build-dirs: instead of git diff --name-only ... 

So, it is the way we want to use this component as is now? Or, are you planning more changes?

matthewhughes-uw replied:
> So, it is the way we want to use this component as is now? Or, are you planning more changes?

The plan is to use it as-is (see updated description)

>  It would also be worth to include information about your planning next steps- how are you finally planning to use this piece of work?

Go shout, updated description

MarcinGinszt commented on cmd/get-kustomize-build-base/README.md:
```diff
@@ -0,0 +1,24 @@
+# `get-kustomize-build-base`
```
Why do we need additional command instead of updating `kustomize-build-dirs`? 
It corresponds to the logic from `kustomize-build-dirs` so much, that it needs function copied from there (findKustomizationRoot). 
It's hard to find the use case  for `get-kustomize-build-base` other than chained with `kustomize-build-dirs` - and it's hard to find the use case for `kustomize-build-dirs` to run without this command.

matthewhughes-uw replied:
> Why do we need additional command instead of updating `kustomize-build-dirs`? It corresponds to the logic from `kustomize-build-dirs` so much, that it needs function copied from there (findKustomizationRoot). It's hard to find the use case for `get-kustomize-build-base` other than chained with `kustomize-build-dirs` - and it's hard to find the use case for `kustomize-build-dirs` to run without this command.

Reasons for a separate command:

* It's simpler: just adding a bunch of code with clearly defined functionality vs. injecting extra functionality into an existing command (though we could move things around in the future if needed)
* It separates intent: `kustomize-build-dirs` can only worry about running `kustomize build` and this command can worry about finding where to build

> so much, that it needs function copied from there (findKustomizationRoot).

Yeah, ideally I will replace that logic with this logic, but I wanted to see if it works well first

Do you have any concerns with a separate command?

MarcinGinszt replied:
> Do you have any concerns with a separate command?

Yes, I believe it greatly increases the cognitive load. 
Do you believe, that making this (already complex command)
 ```
git diff --name-only --raw HEAD^1 HEAD \
            | xargs kustomize-build-dirs --out-dir built-manifests/ --truncate-secrets 2>&1 \
            | tee build-output
```
even more complex, and demanding knowledge of additional component:
```
git diff --name-only --raw HEAD^1 HEAD \
            | get-kustomize-build-base 
            | xargs kustomize-build-dirs --out-dir built-manifests/ --truncate-secrets 2>&1 \
            | tee build-output
```
will simplify the life of the end user?

Or, that creating additional command for sake of running one additional function, which is needed only as an output to another command- will simplify the maintenance of this repository?

MarcinGinszt replied:
>It separates intent: kustomize-build-dirs can only worry about running kustomize build and this command can worry about finding where to build

According to this logic, we could propose- for example- splitting the `kustomize-build-dirs` into [following components](https://github.com/utilitywarehouse/manifest-checkers/blob/main/cmd/kustomize-build-dirs/main.go#L62-L99):

- check-kustomize-installed
- find-kustomization-roots
-  remove-component-kustomization

matthewhughes-uw replied:
Adding it to `kustomize-build-dirs` is non-trivial, since that command expects a bunch of filenames passed as args, which won't work for the output of `git diff --raw`. Since there's a bunch of spaces in each line, I'm not sure we would even know when to start/stop parsing a given line.

My plan is to:

* Slot this command in, see if it really does what we want/gives devs nicer diffs, and if it does
* Come back and tidy up/rethink (e.g. see if we want to throw it all in `kustomize-build-dirs`, and if so make it handle inputs differently etc.)

MarcinGinszt replied:
> that command expects a bunch of filenames passed as args, which won't work for the output of git diff --raw. Since there's a bunch of spaces in each line, I'm not sure we would even know when to start/stop parsing a given line.

Aren't you splitting the result of `git diff --raw` in [this PR](https://github.com/utilitywarehouse/manifest-checkers/pull/51/files#diff-5581e059a643c28d8423489ccb67a131431c91fe004b593c19b071de68110a26R70)?

MarcinGinszt replied:
>Come back and tidy up/rethink (e.g. see if we want to throw it all in kustomize-build-dirs, and if so make it handle inputs differently etc.)

This is customer- facing tool used in the multiple repositories used by different teams. I think we can't afford half- solutions and experiments for this one. 

If there are cases, where we should keep the high quality of code- it's one of them.

matthewhughes-uw replied:
> Aren't you splitting the result of git diff --raw in [this PR](https://github.com/utilitywarehouse/manifest-checkers/pull/51/files#diff-5581e059a643c28d8423489ccb67a131431c91fe004b593c19b071de68110a26R70)?

Yeah, I can do that because this command reads from `stdin`, `kustomize-build-dirs` does not and I would like to avoid changing how it takes inputs+adding extra logic in a single change

MarcinGinszt replied:
> I would like to avoid changing how it takes inputs+adding extra logic in a single change

Obviously, that's understandable. So, what's your plan?

MarcinGinszt replied:
>   Add this new command
    Inject it into the k8s-manifests pipeline as outline above
    Get feedback
    If feedback good, consider merging the logic back into one command or otherwise packaging it nicely so other teams/users can benefit


What are the benefits of this approach?
The only benefit of this approach  which I see, is that we would push the effort into the future. 
It would be simplify the work now, but we would create tech debt*, with hope that "we'll eventually fix it after it will run some time".

*we would have two ways of running the code. And we would have some code replicated in `manifest-checkers`. And we would have mention in the code: "todo: eventually refactor".

MarcinGinszt replied:
>    Code quality: should always be high, things is code readability, having readable tests etc.
    Feature quality: how perfect of a product we present to the users, can vary at times (for experiments etc.)

What's the benefit of nice, clean code, if it is doing weird things? 
I think the logic should be the priority before aesthetics

MarcinGinszt commented on cmd/get-kustomize-build-base/main.go:
```diff
@@ -0,0 +1,141 @@
+
+func getBaseForChange(buildRoot string, status string, files []string) (string, error) {
+	switch status {
+	case changeStatusModification, changeStatusChangingType:
+		return findKustomizationRoot(buildRoot, files[0])
+	case changeStatusAddition, changeStatusDeletion:
+		return findKustomizationRootForNew(buildRoot, files[0])
```
So, if we are adding/ deleting/ copying/ renaming Kustomization, then we look for the `parent` Kustomization file (I suppose, to see if it was included there).
But, if we are modifying the Kustomization, then we don't look for the parent- we take just that. 

Why shouldn't we always look for the `parent`?

MarcinGinszt replied:
Also, currently, if somebody adds parent level kustomization, this tool won't find any changes.

MarcinGinszt replied:
Maybe to solve the problem for all the cases, we could just revert the algorthim for `findKustomizationRoot` - instead of going down the path, we could go up the path from repository level. 
This would mean, we don't need whole of that algorithm to parse the git change type. 
WDYT?

matthewhughes-uw replied:
> Maybe to solve the problem for all the cases, we could just revert the algorthim for findKustomizationRoot - instead of going down the path, we could go up the path from repository level.

I'm not sure I follow, could you expand on what you mean?

MarcinGinszt replied:
If the change now is at `a/b/c/d/file.txt`, we look for kustomization.yaml in
`a/b/c/d`
`a/b/c`
`a/b`
`a`

What if we would look for it in
`a`
`a/b`
`a/b/c`
`a/b/c/d`
?

We would always find the first parent then, and we capture all the changes.

MarcinGinszt replied:
I assume, if this would be correct, then we would just need to modify the `findKustomizationRoot` function in `kustomize-build-dirs` to resolve the issue

MarcinGinszt replied:
This demands assumption, that each kustomization.yaml above in hierarchy is parent- but we operate with this assumption now

matthewhughes-uw replied:
> If the change now is at `a/b/c/d/file.txt`, we look for kustomization.yaml in `a/b/c/d` `a/b/c` `a/b` `a`
> 
> What if we would look for it in `a` `a/b` `a/b/c` `a/b/c/d` ?
> 
> We would always find the first parent then, and we capture all the changes.

:thinking: in this case I think we'll be exclusively building entire namespaces on each change/run. I expect this would slow things down since we'd need to do a bunch of cloning to fetch remove kustomizations that we don't actually 'need' (i.e. from building all the non-changed services in that namespace). Here's a run rebuilding all of (I think one of the largest?) namespaces https://github.com/utilitywarehouse/kubernetes-manifests/actions/runs/10616576962/job/29427336227 it took ~3 minutes. For other namespaces though this should be much quicker. It's probably worth experimenting with

MarcinGinszt commented:
I asked on slack- why were we excluding the `deleted` files from original `build-kustomize-bases` command.

matthewhughes-uw commented:
> I asked on slack- why were we excluding the `deleted` files from original `build-kustomize-bases` command.

copying my response:

depends on the use case:

* If you delete some file that isn't kustomization.yaml and update kustomization.yaml to mark that removal: no change from the current behaviour
* If you delete some file that isn't kustomization.yaml and don't update kustomization.yaml to mark that removal:
    * current setup: on feature-branch we build nothing, on master we build the project. Diff is the entire project
    * if we removed the filter: on feature-branch we would build the project, on master we would build the project
* If you delete some kustomization.yaml:
    * current setup: on master we will build that entire project and diff: the diff will show the entire project as removed
    * If we removed the filter: on feature-branch we would build the parent (since we can't find kustomization.yaml, so we look up one directory) on master we would build just the project. So the diff would look like we added the entire parent project
